src/sentry/utils/performance_issues/performance_detection.py

```diff
@@ -553,11 +564,185 @@ def visit_span(self, span: Span):
+
+    def _contains_complete_query(self, span: Span) -> bool:
+        # When relay truncates span description, it adds a "..." suffix (three
+        # full stops, not an ellipsis).
+        query = span.get("description", None)
+        return query and not query.endswith("...")
+
+    def _maybe_use_as_source(self, span: Span):
+        if not self._contains_complete_query(span):
+            return
+
+        parent_span_id = span.get("parent_span_id", None)
+        if not parent_span_id or parent_span_id not in self.potential_parents:
+            return
+
+        self.source_span = span
+
+    def _continues_n_plus_1(self, span: Span):
+        if not self._contains_complete_query(span):
+            return False
+
+        if self._overlaps_last_span(span):
+            return False
+
+        expected_parent_id = self.source_span.get("parent_span_id", None)
+        parent_id = span.get("parent_span_id", None)
+        if not parent_id or parent_id != expected_parent_id:
+            return False
+
+        span_hash = span.get("hash", None)
+        if not span_hash:
+            return False
+
+        if not self.n_hash:
+            self.n_hash = span_hash
+            return True
+
+        return span_hash == self.n_hash
+
+    def _overlaps_last_span(self, span: Span) -> bool:
+        last_span = self.source_span
+        if self.n_spans:
+            last_span = self.n_spans[-1]
+
+        last_span_ends = timedelta(seconds=last_span.get("timestamp", 0))
+        current_span_begins = timedelta(seconds=span.get("start_timestamp", 0))
+        return last_span_ends > current_span_begins
+
+    def _maybe_store_issue(self):
+        if not self.source_span or not self.n_spans:
+            return
+
+        count = self.settings.get("count")
+        duration_threshold = timedelta(milliseconds=self.settings.get("duration_threshold"))
+
+        # Do we have enough spans?
+        if len(self.n_spans) < count:
+            return
+
+        # Do the spans take enough total time?
+        total_duration = timedelta()
+        for span in self.n_spans:
+            total_duration += get_span_duration(span)
+        if total_duration < duration_threshold:
+            return
+
+        # We require a parent span in order to improve our fingerprint accuracy.
+        parent_span_id = self.source_span.get("parent_span_id", None)
+        if not parent_span_id:
+            return
+        parent_span = self.potential_parents[parent_span_id]
+        if not parent_span:
+            return
+
+        fingerprint = self._fingerprint(
+            parent_span.get("op", None),
+            parent_span.get("hash", None),
+            self.source_span.get("hash", None),
+            self.n_spans[0].get("hash", None),
+        )
+        if fingerprint not in self.stored_issues:
+            self.stored_issues[fingerprint] = PerformanceSpanProblem(
+                fingerprint=fingerprint,
+                spans_involved=self.n_spans,
```

Comment:
Could we also include the source span as one of the spans in `spans_involved`? It will be useful for the embedded span tree, so we can show the entire structure of the N+1 in events

Reply:
100%, good catch. Thanks! 😄

Reply:
Forgot to mention earlier but we should add the source span too, in case you haven't yet 😃